servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/HttpTransportObserverAsyncContextTest.java

```diff
@@ -0,0 +1,328 @@
+    }
+
+    private void testRequestResponse(StreamingHttpRequest request, HttpResponseStatus expectedStatus,
+                                     int expectedResponseLength) throws Exception {
+        StreamingHttpResponse response = streamingHttpClient().request(request).toFuture().get();
+        assertResponse(response, protocol.version, expectedStatus, expectedResponseLength);
+        assertMap(clientObserver.storageMap(), CLIENT_VALUE);
+    }
+
+    private void testServerReadError(String path) throws Exception {
+        StreamingHttpResponse response = streamingHttpClient().request(newRequest(path))
+                .toFuture().get();
+        assertResponse(response, protocol.version, OK);
+        assertThrows(ExecutionException.class, () -> response.payloadBody().ignoreElements().toFuture().get());
+        assertMap(clientObserver.storageMap(), CLIENT_VALUE);
+    }
+
+    private static void assertMap(Map<String, String> map, String expected) {
+        Set<Map.Entry<String, String>> entries = map.entrySet();
+        assertThat("No entries in the map", entries, hasSize(greaterThan(0)));
+        for (Map.Entry<String, String> entry : entries) {
+            assertThat("Unexpected value for \"" + entry.getKey() + "\" callback",
+                    entry.getValue(), equalTo(expected));
+        }
+    }
+
+    private static final class AsyncContextCaptureTransportObserver implements TransportObserver {
+
+        private final Map<String, String> storageMap = new ConcurrentSkipListMap<>();
+        private final AsyncContextMap.Key<String> key;
+
+        AsyncContextCaptureTransportObserver(AsyncContextMap.Key<String> key) {
+            this.key = key;
+        }
+
+        Map<String, String> storageMap() {
+            return unmodifiableMap(storageMap);
+        }
+
+        @Override
+        public ConnectionObserver onNewConnection() {
+            storageMap.put("onNewConnection", valueOf(AsyncContext.get(key)));
+            return new AsyncContextCaptureConnectionObserver();
+        }
+
+        private class AsyncContextCaptureConnectionObserver implements ConnectionObserver {
+
+            @Override
+            public void onDataRead(final int size) {
+                // AsyncContext is unknown at this point because this event is triggered by network
+            }
+
+            @Override
+            public void onDataWrite(final int size) {
+                // AsyncContext is unknown at this point because protocols can write multiple requests concurrently
+            }
+
+            @Override
+            public void onFlush() {
+                // AsyncContext is unknown at this point because protocols can flush multiple requests concurrently
+            }
+
+            @Override
+            public SecurityHandshakeObserver onSecurityHandshake() {
+                // AsyncContext is unknown at this point because this event is triggered by network
```

Comment:
In some cases, `ConnectionObserver` events are generated by network events when the request-related `AsyncContext` is unknown. For those cases, users should capture the `AsyncContext` state from other callbacks, keep it in local variable(s), and propagate when necessary. Examples:
1. `ConnectionObserver#onSecurityHandshake()` and all `SecurityHandshakeObserver` callbacks should use the state captured in `TransportObserver#onNewConnection()`.
2. `ReadObserver`/`WriteObserver` events should use the state captured in `DataObserver` callbacks.
3. `StreamObserver#streamClosed` events should use the state captured in `StreamObserver#streamEstablished()` or `MultiplexedObserver#onNewStream()`.
4. `connectionClosed` events can be correlated with request(s) using `requestId<->connectionId` correlation at the higher levels.

@Scottmitch we can cover more cases (for example, `onSecurityHandshake `) using another technique if necessary. Because we have access to `subscriber` at the time we invoke this callback, we can check if it's an instance of `ContextPreservingSubscriber` and extract the `AsyncContextMap` from it. However, it will require some movements for `ContextPreserving*` classes. I will defer that until we have a use-case.

Reply:
agreed lets defer until necessary.